src/System.Net.WebHeaderCollection/src/System/Net/WebHeaderCollection.cs

```diff
@@ -16,19 +19,38 @@ internal enum WebHeaderCollectionType : byte
         WebResponse,
     }
 
-    public sealed class WebHeaderCollection : IEnumerable
+    [Serializable]
+    public class WebHeaderCollection : NameValueCollection, ISerializable
     {
         private const int ApproxAveHeaderLineSize = 30;
         private const int ApproxHighAvgNumHeaders = 16;
 
-        // Lazily initialized fields.
-        private List<string> _entriesList;
-        private Dictionary<string, string> _entriesDictionary;
-        private string[] _allKeys;
+        // Lazily initialized field.
+        private NameValueCollection _innerCollection;
+        private NameValueCollection InnerCollection
+        {
+            get
+            {
+                if (_innerCollection == null)
+                    _innerCollection = new NameValueCollection(ApproxHighAvgNumHeaders, CaseInsensitiveAscii.StaticInstance);
+                return _innerCollection;
+            }
+        }
 
         // This is the object that created the header collection.
         private WebHeaderCollectionType _type;
 
+        protected WebHeaderCollection(SerializationInfo serializationInfo, StreamingContext streamingContext)
+        {
+            int count = serializationInfo.GetInt32("Count");
```

Comment:
@tijoytom There are some differences from .Net Desktop here. 
Where possible, we prefer identical implementations to ensure the best app-compat behavior between the implementation.
We would also like to have the same logging in both implementation. That helps us debug potential behavior differences reported in customer environments.